internal/controller/boxedjob_controller.go

```diff
@@ -0,0 +1,273 @@
+
+	case workloadv1alpha1.BoxedJobSuspended: // no components deployed
+		if job.Spec.Suspend {
+			return ctrl.Result{}, nil // remain suspended
+		}
+		return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobDeploying) // begin deployment
+
+	case workloadv1alpha1.BoxedJobDeploying: // deploying components
+		if job.Spec.Suspend {
+			return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobSuspending) // abort deployment
+		}
+		err, fatal := r.createComponents(ctx, job)
+		if err != nil {
+			if fatal {
+				return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobDeleting) // abort on fatal error
+			}
+			return ctrl.Result{}, err // retry creation on transient error
+		}
+		return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobRunning)
+
+	case workloadv1alpha1.BoxedJobRunning: // components deployed
+		if job.Spec.Suspend {
+			return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobSuspending) // begin undeployment
+		}
+		completed, err := r.hasCompleted(ctx, job)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if completed {
+			return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobCompleted)
+		}
+		failed, err := r.hasFailed(ctx, job)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if failed {
+			return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobDeleting)
+		}
+		return ctrl.Result{RequeueAfter: time.Minute}, nil
+
+	case workloadv1alpha1.BoxedJobSuspending: // undeploying components
+		// finish undeploying components irrespective of desired state (suspend bit)
+		if r.deleteComponents(ctx, job) {
+			return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobSuspended)
+		} else {
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+	case workloadv1alpha1.BoxedJobDeleting: // deleting components on failure
+		if r.deleteComponents(ctx, job) {
+			return r.updateStatus(ctx, job, workloadv1alpha1.BoxedJobFailed)
+		} else {
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *BoxedJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&workloadv1alpha1.BoxedJob{}).
+		Watches(&v1.Pod{}, handler.EnqueueRequestsFromMapFunc(r.podMapFunc)).
+		Complete(r)
+}
+
+func (r *BoxedJobReconciler) podMapFunc(ctx context.Context, obj client.Object) []reconcile.Request {
+	pod := obj.(*v1.Pod)
+	if name, ok := pod.Labels[boxedJobLabel]; ok {
+		if pod.Status.Phase == v1.PodSucceeded {
+			return []reconcile.Request{{NamespacedName: types.NamespacedName{Namespace: pod.Namespace, Name: name}}}
+		}
+	}
+	return nil
+}
+
+func parseComponent(job *workloadv1alpha1.BoxedJob, raw []byte) (*unstructured.Unstructured, error) {
+	obj := &unstructured.Unstructured{}
+	if _, _, err := unstructured.UnstructuredJSONScheme.Decode(raw, nil, obj); err != nil {
+		return nil, err
+	}
+	// TODO: use heuristic to inject labels
```

Comment:
I think with the change to `path` we don't need heuristics.  We can piggyback on the podset label injection we have to implement for kueue.

Reply:
agreed